File: data_structure/链表/c/DoubleLinkedList.c
#include <stdio.h>
#include <stdlib.h>

// 双链表结点
typedef struct Node {
  int data;
  struct Node *prev;
  struct Node *next;
} Node;

// 双链表的结构
typedef struct DoubleLinkedList {
  Node *head;
  Node *tail;
  int size;
  int capacity;
} DoubleLinkedList;

// 创建双链表
DoubleLinkedList *createDoubleLinkedList(int capacity) {
  if (capacity <= 0) {
    printf("链表的容量不能为 0。\n");
    return NULL;
  }

  DoubleLinkedList *list = (DoubleLinkedList *)malloc(sizeof(DoubleLinkedList));

  if (!list) {
    printf("创建双链表时，申请内存失败。\n");
    return NULL;
  }

  list->head = NULL;
  list->tail = NULL;
  list->size = 0;
  list->capacity = capacity;
  return list;
}

// 查询双链表的容量
int getCapacity(DoubleLinkedList *list) { return list->capacity; }

// 查询双链表当前的长度
int getSize(DoubleLinkedList *list) { return list->size; }

// 判断当前链表是否为空
int isEmpty(DoubleLinkedList *list) { return list->size == 0; }

// 判断当前链表是否满了
int isFull(DoubleLinkedList *list) { return list->size == list->capacity; }

// 获取链表的头结点
Node *getHeadNode(DoubleLinkedList *list) {
  if (isEmpty(list)) {
    printf("当前链表为空。\n");
    return NULL;
  }
  return list->head;
}

// 获取链表的尾结点
Node *getTailNode(DoubleLinkedList *list) {
  if (isEmpty(list)) {
    printf("当前链表为空。\n");
    return NULL;
  }
  return list->tail;
}

// 从头结点添加元素
int addNodeFromHead(DoubleLinkedList *list, int data) {
  if (isFull(list)) {
    printf("从头结点添加元素，当前链表已满，不能添加新元素。\n");
    return 0;
  }

  Node *newNode = (Node *)malloc(sizeof(Node));
  if (!newNode) {
    printf("从头结点添加元素，创建新结点时，申请内存失败。\n");
    return 0;
  }

  newNode->data = data;
  newNode->prev = NULL;
  newNode->next = list->head;

  if (list->head != NULL) {
    list->head->prev = newNode;
  }
  list->head = newNode;

  if (list->tail == NULL) {
    list->tail = newNode;
  }
  list->size++;
  return 1;
}

// 从尾结点添加元素
int addNodeFromTail(DoubleLinkedList *list, int data) {
  if (isFull(list)) {
    printf("从尾结点添加元素，当前链表已满，不能添加新元素。\n");
    return 0;
  }

  Node *newNode = (Node *)malloc(sizeof(Node));
  if (!newNode) {
    printf("从尾结点添加元素，创建新结点时，申请内存失败。\n");
    return 0;
  }

  newNode->data = data;
  newNode->prev = list->tail;
  newNode->next = NULL;

  if (list->tail != NULL) {
    list->tail->next = newNode;
  }
  list->tail = newNode;

  if (list->head == NULL) {
    list->head = newNode;
  }

  list->size++;
  return 1;
}

// 遍历整个链表，并输出
void traverseList(DoubleLinkedList *list) {
  if (list == NULL || list->head == NULL) {
    printf("当前链表为空，无需遍历。\n");
    return;
  }
  Node *currentNode = list->head;
  printf("开始遍历链表：[ ");
  while (currentNode != NULL) {
    printf("%d ", currentNode->data);
    currentNode = currentNode->next;
  }

  printf("] 长度: %d\n", list->size);
}

int main() {
  DoubleLinkedList *list = createDoubleLinkedList(5);
  addNodeFromTail(list, 1);
  addNodeFromTail(list, 2);
  addNodeFromTail(list, 3);
  addNodeFromTail(list, 4);
  addNodeFromTail(list, 5);
  addNodeFromTail(list, 6);
  traverseList(list);
  printf("\n");
}

File: data_structure/链表/java/DoubleLinkedList.java
class DoubleNode {
  public int data;
  DoubleNode prev;
  DoubleNode next;

  public DoubleNode(int data) {
    this.data = data;
    this.prev = null;
    this.next = null;
  }

}

public class DoubleLinkedList {

  public DoubleNode head;
  public DoubleNode tail;
  public int size;
  public int capacity;

  public DoubleLinkedList(int capacity) {
    if (capacity <= 0) {
      System.out.println("双链表的容量不能为空");
      this.capacity = 0;
      this.size = 0;
      this.head = null;
      this.tail = null;
      return;
    }
    this.capacity = capacity;
    this.size = 0;
    this.head = null;
    this.tail = null;
  }

  int getSize() {
    return this.size;
  }

  int getCapacity() {
    return this.capacity;
  }

  boolean isEmpty() {
    return this.size == 0;
  }

  boolean isFull() {
    return this.capacity == this.size;
  }

  DoubleNode getHeadNode() {
    return this.head;
  }

  DoubleNode getTailNode() {
    return this.tail;
  }

  // 从头节点添加元素
  int addNodeFromHead(int data) {
    if (this.isFull()) {
      System.out.println("从头节点添加元素，当前链表已满，不能添加新元素。");
      return 0;
    }

    DoubleNode newNode = new DoubleNode(data);
    newNode.prev = null;
    newNode.next = this.head;

    if (this.head != null) {
      this.head.prev = newNode;
    }
    this.head = newNode;

    if (this.tail == null) {
      this.tail = newNode;
    }

    this.size++;
    return 1;
  }

  // 从尾节点添加元素
  int addNodeFromTail(int data) {
    if (this.isFull()) {
     System.out.println("从尾节点添加元素，当前链表已满，不能添加新元素。");
      return 0;
    }

    DoubleNode newNode = new DoubleNode(data);
    newNode.prev = this.tail;
    newNode.next = null;

    if (this.tail != null) {
      this.tail.next = newNode;
    }
    this.tail = newNode;

    if (this.head == null) {
      this.head = newNode;
    }

    this.size++;
    return 1;
  }

  // 删除节点
  int removeNode(int data) {
    if (this.isEmpty()) {
      System.out.println("当前链表为空，无需遍历。");
      return 0;
    }

    DoubleNode currentNode = this.head;

    while (currentNode != null && currentNode.data != data) {
      currentNode = currentNode.next;
    }

    if (currentNode == null) {
      System.out.printf("未找到值为 %d 的节点，删除失败。\n", data);
      return 0;
    }

    // 说明是首结点
    if (currentNode.prev == null) {
      this.head = currentNode.next;
    } else {
      currentNode.prev.next = currentNode.next;
    }

    if (currentNode.next == null) {
      this.tail = currentNode.prev;
    } else {
      currentNode.next.prev = currentNode.prev;
    }

    this.size--;
    return 1;
  }

  // 查找节点
  DoubleNode findNode(int data) {
    if (this.isEmpty()) {
      System.out.println("当前链表为空，不能查找元素。");
      return null;
    }

    DoubleNode currentNode = this.head;
    while (currentNode != null) {
      if (currentNode.data == data) {
        return currentNode;
      }
      currentNode = currentNode.next;
    }
    return null;
  }


  // 遍历链表
  void traverse() {
    if (this.isEmpty()) {
      System.out.println("当前链表为空，无需遍历。");
      return;
    }

    DoubleNode currentNode = this.head;
    System.out.printf("开始遍历链表：[ ");
    while (currentNode != null) {
      System.out.printf("%d ", currentNode.data);
      currentNode = currentNode.next;
    }
    System.out.printf("] 长度: %d\n", this.size);
  }

}
